lib/ansible/modules/network/routeros/routeros_facts.py

```diff
@@ -0,0 +1,434 @@
+        preprocessed = list()
+        for line in data.split('\n'):
+            if len(line) == 0 or line[:5] == 'Flags':
+                continue
+            elif re.match(r'\s\d', line[:2]):
+                preprocessed.append(line[2:])
+            else:
+                preprocessed[-1] += line
+        return preprocessed
+
+    def parse_interfaces(self, data):
+        facts = dict()
+        data = self.preprocess(data)
+        for line in data:
+            name = self.parse_name(line)
+            facts[name] = dict()
+            for (key, value) in re.findall(r'([\w\d\-]+)=\"?(\w{3}/\d{2}/\d{4}\s\d{2}:\d{2}:\d{2}|[\w\d\-\.:/]+)', line):
+                facts[name][key] = value
+        return facts
+
+    def parse_addresses(self, data):
+        facts = dict()
+        data = self.preprocess(data)
+        for line in data:
+            name = self.parse_interface(line)
+            facts[name] = dict()
+            for (key, value) in re.findall(r'([\w\d\-]+)=\"?(\w{3}/\d{2}/\d{4}\s\d{2}:\d{2}:\d{2}|[\w\d\-\.:/]+)', line):
+                facts[name][key] = value
+        return facts
+
+    def parse_neighbors(self, data):
+        facts = dict()
+        data = self.preprocess(data)
+        for line in data:
+            name = self.parse_interface(line)
+            facts[name] = dict()
+            for (key, value) in re.findall(r'([\w\d\-]+)=\"?(\w{3}/\d{2}/\d{4}\s\d{2}:\d{2}:\d{2}|[\w\d\-\.:/]+)', line):
+                facts[name][key] = value
+        return facts
+
+    def parse_name(self, data):
+        match = re.search(r'name=\"([\w\d\-]+)\"', data, re.M)
+        if match:
+            return match.group(1)
+
+    def parse_interface(self, data):
+        match = re.search(r'interface=([\w\d\-]+)', data, re.M)
+        if match:
+            return match.group(1)
+
+
+FACT_SUBSETS = dict(
+    default=Default,
+    hardware=Hardware,
+    interfaces=Interfaces,
+    config=Config,
+)
+
+VALID_SUBSETS = frozenset(FACT_SUBSETS.keys())
+
+global warnings
+warnings = list()
+
+
+def main():
+    """main entry point for module execution
+    """
+    argument_spec = dict(
+        gather_subset=dict(default=['!config'], type='list')
+    )
+
+    argument_spec.update(routeros_argument_spec)
+
+    module = AnsibleModule(argument_spec=argument_spec,
+                           supports_check_mode=True)
+
+    gather_subset = module.params['gather_subset']
+
+    runable_subsets = set()
+    exclude_subsets = set()
+
+    for subset in gather_subset:
+        if subset == 'all':
+            runable_subsets.update(VALID_SUBSETS)
+            continue
+
+        if subset.startswith('!'):
+            subset = subset[1:]
+            if subset == 'all':
+                exclude_subsets.update(VALID_SUBSETS)
+                continue
+            exclude = True
+        else:
+            exclude = False
+
+        if subset not in VALID_SUBSETS:
+            module.fail_json(msg='Bad subset')
```

Comment:
Maybe output the bad value here as well, to help the user find out which one was wrong.